gen/src/main/java/com/megaease/easeagent/gen/AdviceProcessor.java

```diff
@@ -0,0 +1,103 @@
+ * limitations under the License.
+ */
+package com.megaease.easeagent.gen;
+
+import com.google.auto.service.AutoService;
+
+import javax.annotation.processing.AbstractProcessor;
+import javax.annotation.processing.Filer;
+import javax.annotation.processing.Processor;
+import javax.annotation.processing.RoundEnvironment;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.util.Elements;
+import javax.tools.Diagnostic.Kind;
+import javax.tools.FileObject;
+import javax.tools.StandardLocation;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+import static java.util.Collections.singleton;
+import static javax.lang.model.element.Modifier.ABSTRACT;
+
+@AutoService(Processor.class)
+public class AdviceProcessor extends AbstractProcessor {
+    Class<Generate.Advice> annotationClass = Generate.Advice.class;
+
+    @Override public SourceVersion getSupportedSourceVersion() {
+        return SourceVersion.latestSupported();
+    }
+
+    @Override
+    public Set<String> getSupportedAnnotationTypes() {
+        return singleton(annotationClass.getCanonicalName());
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+        if (roundEnv.processingOver()) {
+            return false;
+        }
+
+        Map<String,Set<String>> services = new HashMap<>();
+        String cn = "com.megaease.easeagent.core.Transformation";
+        Elements elements = processingEnv.getElementUtils();
+
+        Set<? extends Element> roundElements = roundEnv.getElementsAnnotatedWith(annotationClass);
+        for (Element e : roundElements) {
+            if (!e.getKind().isClass()) {
+                continue;
+            }
+            if (!e.getModifiers().contains(ABSTRACT)) {
+                continue;
+            }
+            TypeElement type = (TypeElement)e;
+
+            Set<String> v = services.computeIfAbsent(cn, k -> new TreeSet<>());
+            v.add(elements.getBinaryName(type).toString());
+        }
+
+        Filer filer = processingEnv.getFiler();
+        String prefix = "META-INF/services/";
+
+        Set<String> providers = services.get(cn);
+        if (providers == null || providers.isEmpty()) {
+            return false;
+        } else {
+            try {
+                String contract = prefix + cn;
+                processingEnv.getMessager().printMessage(Kind.NOTE,"Writing " + contract);
+                FileObject f = filer.createResource(StandardLocation.CLASS_OUTPUT, "", contract);
+                PrintWriter pw = new PrintWriter(new OutputStreamWriter(f.openOutputStream(), StandardCharsets.UTF_8));
+                providers.forEach(pw::println);
+                pw.close();
```

Comment:
The PrintWriter should be closed in finally block

Reply:
Done